core/src/main/scala/cats/data/NonEmptyList.scala

```diff
@@ -0,0 +1,193 @@
+package cats
+package data
+
+import cats.instances.list._
+import cats.syntax.eq._
+
+import scala.annotation.tailrec
+import scala.collection.mutable.ListBuffer
+
+/**
+ * A data type which represents a non empty list of A, with
+ * single element (head) and optional structure (tail).
+ */
+final case class NonEmptyList[A](head: A, tail: List[A]) {
```

Comment:
what if we did something similar to `NonEmptyVector` and made this a value class around `::[A]`? Then we would not need to allocate in some cases.